compiler/rustc_middle/src/mir/interpret/mod.rs

```diff
@@ -554,6 +641,129 @@ impl<'tcx> TyCtxt<'tcx> {
     fn set_alloc_id_same_memory(self, id: AllocId, mem: &'tcx Allocation) {
         self.alloc_map.lock().alloc_map.insert_same(id, GlobalAlloc::Memory(mem));
     }
+
+    /// Store the result of a call to `eval_to_allocation_raw` in order to
+    /// allow deduplication.
+    #[instrument(skip(self), level = "debug")]
+    pub fn save_alloc_for_dedup(self, id: GlobalId<'tcx>, val: ConstDedupResult<ConstAlloc<'tcx>>) {
+        let dedup_const_map = self.dedup_const_map.lock();
+        dedup_const_map.insert_alloc(id, val);
+        debug!("dedup_const_map after insert: {:#?}", dedup_const_map);
+    }
+
+    /// Store the result of a call to `eval_to_const_value_raw` in order to deduplicate it.
+    #[instrument(skip(self), level = "debug")]
+    pub fn save_const_value_for_dedup(
+        self,
+        id: GlobalId<'tcx>,
+        val: ConstDedupResult<ConstValue<'tcx>>,
+    ) {
+        let dedup_const_map = self.dedup_const_map.lock();
+        dedup_const_map.insert_const_val(id, val);
+        debug!("dedup_const_map after insert: {:#?}", dedup_const_map);
+    }
+
+    /// Tries to deduplicate a call to `eval_to_allocation_raw`. If deduplication isn't
+    /// successful `eval_to_allocation_raw` query is executed.
+    #[instrument(skip(self, opt_span), level = "debug")]
+    pub fn dedup_eval_alloc_raw(
+        self,
+        key: ty::ParamEnvAnd<'tcx, GlobalId<'tcx>>,
+        opt_span: Option<Span>,
+    ) -> EvalToAllocationRawResult<'tcx> {
+        use ConstDedupResult::*;
+
+        let (param_env, id) = key.into_parts();
+        let dedup_const_map = self.dedup_const_map.lock();
+        debug!("dedup_const_map: {:#?}", dedup_const_map);
+        let alloc_map = dedup_const_map.alloc_map.borrow();
+        debug!("alloc_map: {:#?}", alloc_map);
+
+        let dedup_result = alloc_map.get(&id);
+        debug!(?dedup_result);
+
+        match param_env.reveal() {
```

Comment:
these matches are not obvious to me. Can you explain them? I'm unsure what the rules are here.

Reply:
The rationale for those was as follows:

If we have stored a successful allocation with `Reveal::Selection` then we can deduplicate that result on calls of `dedup_eval_alloc_raw` with both `Reveal::Selection` and `Reveal::UserFacing` (since `Selection` and `UserFacing` only differ in that we keep certain errors silent with `Selection`, but successful allocations are always equivalent). This is also the reason why we can deduplicate from saved calls with `UserFacing`on calls with both `Selection` and `UserFacing`. 

If we try to deduplicate with `Reveal::All` we should be able to deduplicate from saved query results with both `Selection` and `UserFacing`, since `All` will always succeed if either `UserFacing` or `Selection` have succeeded. On the other hand if we have a saved result with `Reveal::All`, it's not guaranteed that query calls with `Reveal::Selection` or `Reveal::UserFacing` succeed, so we can't deduplicate from those.

Reply:
👍 please copy this comment into the source, it explains this perfectly